core/chains/evm/txmgr/stuck_tx_detector.go

```diff
@@ -0,0 +1,373 @@
+	}
+	// No transactions found
+	if len(txs) == 0 {
+		return nil, nil
+	}
+
+	switch d.chainType {
+	case config.ChainScroll:
+		return d.detectStuckTransactionsScroll(ctx, txs)
+	case config.ChainZkEvm:
+		return d.detectStuckTransactionsZkEVM(ctx, txs)
+	default:
+		return d.detectStuckTransactionsHeuristic(ctx, txs, blockNum)
+	}
+}
+
+// Finds the lowest nonce Unconfirmed transaction for each enabled address
+// Only the earliest transaction can be considered terminally stuck. All others may be valid and just stuck behind the nonce
+func (d *stuckTxDetector) FindUnconfirmedTxWithLowestNonce(ctx context.Context, enabledAddresses []common.Address) ([]Tx, error) {
+	// Loads attempts within tx
+	txs, err := d.txStore.FindTxsByStateAndFromAddresses(ctx, enabledAddresses, txmgr.TxUnconfirmed, d.chainID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve unconfirmed transactions for enabled addresses: %w", err)
+	}
+	// Stores the lowest nonce tx found in the query results for each from address
+	lowestNonceTxMap := make(map[common.Address]Tx)
+	for _, tx := range txs {
+		if _, ok := lowestNonceTxMap[tx.FromAddress]; !ok {
+			lowestNonceTxMap[tx.FromAddress] = *tx
+		} else if lowestNonceTx := lowestNonceTxMap[tx.FromAddress]; *lowestNonceTx.Sequence > *tx.Sequence {
+			lowestNonceTxMap[tx.FromAddress] = *tx
+		}
+	}
+
+	// Build list of potentially stuck tx but exclude any that are already marked for purge
+	var stuckTxs []Tx
+	for _, tx := range lowestNonceTxMap {
+		// Attempts are loaded newest to oldest so one marked for purge will always be first
+		if len(tx.TxAttempts) > 0 && !tx.TxAttempts[0].IsPurgeAttempt {
+			stuckTxs = append(stuckTxs, tx)
+		}
+	}
+
+	return stuckTxs, nil
+}
+
+// Uses a heuristic to determine a stuck transaction potentially due to overflow
+// This method can be unreliable and may result in false positives but it is best effort to keep the TXM from getting blocked
+// 1. Check if Threshold amount of blocks have passed since the last purge of a tx for the same fromAddress
+// 2. If 1 is true, check if Threshold amount of blocks have passed since the initial broadcast
+// 3. If 2 is true, check if the transaction has at least MinAttempts amount of broadcasted attempts
+// 4. If 3 is true, check if the latest attempt's gas price is higher than what our gas estimator's GetFee method returns
+// 5. If 4 is true, the transaction is likely stuck due to overflow
+func (d *stuckTxDetector) detectStuckTransactionsHeuristic(ctx context.Context, txs []Tx, blockNum int64) ([]Tx, error) {
+	d.purgeBlockNumLock.RLock()
+	defer d.purgeBlockNumLock.RUnlock()
+	// Get gas price from internal gas estimator
+	// Send with arbitrarily high max gas price to prevent the results from being capped. Need the market gas price here.
+	marketGasPrice, _, err := d.gasEstimator.GetFee(ctx, []byte{}, 0, assets.Ether(100))
```

Comment:
The 100 Ether may not be that high a price on some cheap chains using native gas token.
Could you instead take the config's max_price, and multiply it by 2 here?
This way, we are sure we will get uncapped answer.